examples/authentication/client/src/main.rs

```diff
@@ -0,0 +1,203 @@
+        .append_pair("client_id", &opt.client_id);
+    auth_endpoint.to_string()
+}
+
+/// Handles the redirects to extract code from the query string.
+///
+/// A new instance is created for every incoming request. The redirect URL contains the result of
+/// the authentication as query string parameters. If the the authentication is successful the
+/// authentication code is be passed in the `code` paramter. If it failed the reason is passed in
+/// the `error` paramter.
+///
+/// The `result_sender` is used to communicate the results back to the main function.
+///
+/// The handler is responsible for sending an appropriate response to the client browser.
+struct RedirectHandler {
+    result_sender: Sender<Result<String, String>>,
+}
+
+impl Service<Request<Body>> for RedirectHandler {
+    type Response = Response<Body>;
+    type Error = hyper::Error;
+    type Future = future::Ready<Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Ok(()).into()
+    }
+
+    fn call(&mut self, request: Request<Body>) -> Self::Future {
+        if let Some(query) = request.uri().query() {
+            let lookup: HashMap<_, _> = form_urlencoded::parse(query.as_bytes()).collect();
+            if let Some(code) = lookup.get("code") {
+                let code = code.to_string();
+                info!("Auth code: {:?}", code);
+                self.result_sender.try_send(Ok(code)).unwrap();
+                future::ok(Response::new(Body::from("Success!")))
+            } else if let Some(error) = lookup.get("error") {
+                let error = error.to_string();
+                warn!("Error: {:?}", error);
+                self.result_sender.try_send(Err(error)).unwrap();
+                future::ok(
+                    Response::builder()
+                        .status(StatusCode::UNAUTHORIZED)
+                        .body(Body::from("Authentication failed!"))
+                        .unwrap(),
+                )
+            } else {
+                future::ok(
+                    Response::builder()
+                        .status(StatusCode::BAD_REQUEST)
+                        .body(Body::from("Invalid query string!"))
+                        .unwrap(),
+                )
+            }
+        } else {
+            future::ok(
+                Response::builder()
+                    .status(StatusCode::BAD_REQUEST)
+                    .body(Body::from("No query string!"))
+                    .unwrap(),
+            )
+        }
+    }
+}
+
+/// Produces instances of the redirect handler service.
+///
+/// Hyper uses it to create a new handler for every incoming request. The handler needs a copy of
+/// the reference to the ResultSender to communicate the results back to the main thread.
+struct MakeHandler {
+    result_sender: Sender<Result<String, String>>,
+}
+
+impl<T> Service<T> for MakeHandler {
```

Comment:
I find the naming still a bit confusing; is it correct to read `MakeHandler` as handling connections, and `RedirectHandler` as handling requests? If so, maybe it would be more readable to actually replace `T` with `Connection` (or whatever concrete type) (if it works).

Reply:
MakeHandler does not really handle connections, even though a I think connection is passed in to `call` but ignored. It is similar a producer pattern where it is a service responsible for creating one instance of the request handler for each incoming request. I used the naming pattern of beginning the name with `Make` as this seems to be what all the hyper examples do. I have now changed the name to RedirectHandlerProducer as it is clearer about the intention of the service. I don't think it is possible to make it a concrete type, as it could be basically anything that supports `AsyncRead`, `AsyncWrite`. In the case of TLS, it is the TLS stream rather than the original TCP connection that is used. I also think it differs depending on whether the client is using HTTP v1 or v2.